### ISSUE TITLE
Using External Plugin in CoreDNS

### ISSUE THREAD

**Issue by cekbote** [cekbote/coredns_ml_plugin#2]
I am trying to follow this: https://github.com/coredns/example, however I cant see "example" anywhere. 
The starting part of my corefile.5.md file is :
. {
    forward . 9.9.9.9
    example
}

I have changed the file format of the corefile and made it a .txt file. 
It can be found here: [corefile.5.md.txt](https://github.com/Chanakya-Ekbote/coredns_dns_ml_firewall/files/4644961/corefile.5.md.txt)

**Comment by yongtang** [comment]
@Chanakya-Ekbote Is example directory placed under your local coredns repo?

**Comment by cekbote** [comment]
@yongtang Do you mean on my local machine or a forked coredns repo? I can't see anything in my local machine nor I have I forked coredns/coredns. Should I fork it?

**Comment by cekbote** [comment]
@yongtang Follow up on a different external plugin: So what I did was fork the coredns/demo into my repo. Then I added demo:github.com/Chanakya-Ekbote/demo into the plugin.cfg file.  However when I use make, I get the following error:

CGO_ENABLED=0  go build -v -ldflags="-s -w -X github.com/coredns/coredns/coremain.GitCommit=a3aeb3d5-dirty" -o coredns
go: finding github.com/Chanakya-Ekbote/demo latest
go: github.com/coredns/coredns imports
        github.com/coredns/coredns/core/plugin imports
        github.com/Chanakya-Ekbote/demo imports
        github.com/mholt/caddy: github.com/mholt/caddy@v1.0.5: parsing go.mod:
        module declares its path as: github.com/caddyserver/caddy
                but was required as: github.com/mholt/caddy
Makefile:16: recipe for target 'coredns' failed
make: *** [coredns] Error 1

**Comment by yongtang** [comment]
@Chanakya-Ekbote You should build coredns repo first. After a successful build, place example or demo directory under the coredns repo. And make adjustment to build coredns again. This will enable the plugin in coredns (you still need coredns).

**Comment by cekbote** [comment]
@yongtang So you want me to fork the coredns repo. Use that to clone the same forked repo into my local system and then place example or demo directory under the forked repo and clone it again?

**Comment by yongtang** [comment]
@Chanakya-Ekbote You can:
1. Clone coredns in your local disk
2. Build coredns server first
3. If build is successful, place example or demo directory under your local disk's coredns plugin directory, and make plugin.cfg changes.
4. Build coredns server again.

**Comment by cekbote** [comment]
@yongtang Oh I had already cloned CoreDNS and built it. I was going to put the demo directory under my local disks coredns plugin directory, however @greenpau recomended not doing that (Please see gitter). Il do the same now. :)

**Comment by yongtang** [comment]
@Chanakya-Ekbote There are multiple ways to add additional plugin for coredns. Let's start with an short path to get the pipeline. We can come back and revisit once we have a round trip plumbing in place.

**Comment by cekbote** [comment]
@yongtang So I tried using the demo plugin (by creating the folder under the plugin directory and putting the plugins there) as you suggested, and the error is consistent with what I tried earlier:

go generate coredns.go
CGO_ENABLED=0  go build -v -ldflags="-s -w -X github.com/coredns/coredns/coremain.GitCommit=c9c21270-dirty" -o coredns
go: github.com/coredns/coredns imports
        github.com/coredns/coredns/core/plugin imports
        github.com/coredns/coredns/plugin/demo imports
        github.com/mholt/caddy: github.com/mholt/caddy@v1.0.5: parsing go.mod:
        module declares its path as: github.com/caddyserver/caddy
                but was required as: github.com/mholt/caddy
Makefile:16: recipe for target 'coredns' failed
make: *** [coredns] Error 1

**Comment by cekbote** [comment]
@yongtang I tried doing the same with the example folder also. I can see the plugin is enabled. However, I am not getting the output that the link specifies. 
![Plugins](https://user-images.githubusercontent.com/51071680/82245922-73afcd00-9961-11ea-9e5f-1ed0fe564713.PNG)
![Client_Server](https://user-images.githubusercontent.com/51071680/82245925-74486380-9961-11ea-94a5-077321d8dcb2.PNG)
![Corefile](https://user-images.githubusercontent.com/51071680/82245927-75799080-9961-11ea-8416-2bf3073c1656.PNG)

**Comment by greenpau** [comment]
@Chanakya-Ekbote , where is the code you are compiling?

**Comment by cekbote** [comment]
> 
> 
> @Chanakya-Ekbote , where is the code you are compiling?

Its the same code from this link https://github.com/coredns/example
The previous one where I got an error was in: https://github.com/coredns/demo

Looking at these examples I think both the methods (using the github link and putting it under their respective folders in the plugin directory) work well for building plugins.

**Comment by greenpau** [comment]
> Its the same code from this link https://github.com/coredns/example

@Chanakya-Ekbote , please create a pull request with that code so that we could comment on it.

**Comment by cekbote** [comment]
> 
> 
> > Its the same code from this link https://github.com/coredns/example
> 
> @Chanakya-Ekbote , please create a pull request with that code so that we could comment on it.

I did something stupid, I changed the readme file, Il make some small changes in the code too. I should just change the setup.go and the example.go right?

I have just added comments to the files. I hope this is ok. Let me know if anything else is required. :)

**Comment by yongtang** [comment]
@Chanakya-Ekbote let's take a step back. Can you build the coredns without example/demo plugin? Just clone github.com/coredns/coredns locally, and build by following the instructions on README.md from github.com/coredns/coredns

**Comment by cekbote** [comment]
@yongtang @greenpau I actually built coredns locally day before yesterday itself (Please see Gitter, the daily progress doc: https://docs.google.com/document/d/1eD9uu1dLaDtKHJR9HDCYK2CQkRVUnWD0-INHghYb8DE/edit?usp=sharing, also the photos above.)

The reason I created the 'Example' PR, as recommended by @greenpau was that I thought you would comment on it i.e. make changes to the code/or explain it enough so that I could figure it out. Apologies if I  thought wrong.

**Comment by yongtang** [comment]
@Chanakya-Ekbote From the screenshot I see you are showing `corefile.5.md`. Is the content the same as the `Corefile`?

**Comment by cekbote** [comment]
@greenpau @yongtang So from the call I guess it's pretty clear that I am weak in DevOps i.e. git, go. Moreover, I still have no idea how to set up a plugin, nor do I know how to setup a corns server properly. I honestly don't understand most of the things that happened over the call to be brutally honest. The git part I think I can come up to speed within a week or so (by consulting some of my friends and online blogs/videos).  I am really unsure how to proceed from here so could you please give your recomendations?

**Comment by yongtang** [comment]
@Chanakya-Ekbote Yes I think spend some time on git and golang will be worth the effort for this project, as this project is centering around coredns (and to a larger extent cncf/k8s), which are pretty much in golang exclusively.

It is not necessarily to be an expert, but this week I think you want to achieve the following:
1. Build coredns (no external plugin, just coredns itself) from source with golang and be familiar.
2. Familiar with git and golang and see how Golang's import path works.

Once you can build coredns from source with golang, and get familiar with Golang's path/etc, it will be a lot easy to move to adding external plugin.

**Comment by cekbote** [comment]
> 
> 
> @Chanakya-Ekbote Yes I think spend some time on git and golang will be worth the effort for this project, as this project is centering around coredns (and to a larger extent cncf/k8s), which are pretty much in golang exclusively.
> 
> It is not necessarily to be an expert, but this week I think you want to achieve the following:
> 
>     1. Build coredns (no external plugin, just coredns itself) from source with golang and be familiar.
> 
>     2. Familiar with git and golang and see how Golang's import path works.
> 
> 
> Once you can build coredns from source with golang, and get familiar with Golang's path/etc, it will be a lot easy to move to adding external plugin.

Yes I'll do that :)

**Comment by cekbote** [comment]
@yongtang @greenpau So I saw this course online called the missing semester, It talks about using bash, git etc. It can be found here: https://missing.csail.mit.edu/. 

My plan is to do this course until the part where git is explained. Along with the git lecture, I will also go through the Pro Git book https://git-scm.com/book/en/v2. I will do this today and tomorrow. I have already done the first three lecture of the course along with the exercises. (I will finish four today).

From the day after tomorrow to around Monday, I will go through the Learning Go book by Mark and then before the call on Tuesday, I will try to configure coredns onto a Linux based system (The Linux system would be on a VM).

Please let me know your thoughts let me know whether this sounds like a good plan.

**Comment by yongtang** [comment]
@Chanakya-Ekbote An easy way to compile coredns on Linux, without the confusing golang import path, is to build coredns with docker (https://github.com/coredns/coredns#compilation-with-docker)
```
$ docker run --rm -i -t -v $PWD:/v -w /v golang:1.14 make
```

This will give you a repeatable build on linux (the generated binary only works on Linux though. On Windows additional steps are needed.

**Comment by cekbote** [comment]
> 
> 
> @Chanakya-Ekbote An easy way to compile coredns on Linux, without the confusing golang import path, is to build coredns with docker (https://github.com/coredns/coredns#compilation-with-docker)
> 
> ```
> $ docker run --rm -i -t -v $PWD:/v -w /v golang:1.14 make
> ```
> 
> This will give you a repeatable build on linux (the generated binary only works on Linux though. On Windows additional steps are needed.

Sure, il setup a VM and try it. I naturally will have to install docker right? May, I know your thoughts on the plan above :)

**Comment by yongtang** [comment]
Docker is not necessarily for this GSoC project or build coredns itself. Some people prefer to use other methods that may tied to their work environment setup.

Docker is one easy way to build coredns. Since this GSoC project itself is not focusing on building coredns, docker is a method that may help you to avoid spending unnecessary time.

I don't have a strong preference what tool you decide to use, you only need to choose ONE tool which you feel most convenient and easy yourself.

**Comment by cekbote** [comment]
> 
> 
> Docker is not necessarily for this GSoC project or build coredns itself. Some people prefer to use other methods that may tied to their work environment setup.
> 
> Docker is one easy way to build coredns. Since this GSoC project itself is not focusing on building coredns, docker is a method that may help you to avoid spending unnecessary time.
> 
> I don't have a strong preference what tool you decide to use, you only need to choose ONE tool which you feel most convenient and easy yourself.

Sure, Il do that, I am downloading the iso file of ubuntu 18.04. I'll keep you posted on the developments. :)

**Comment by cekbote** [comment]
@yongtang Hi :), I installed a VM and then Ubuntu 18.04. I then installed Docker, and then built coredns. I have attached a screenshot that shows what I am getting. On the right is where I have started the coredns server, on the left is where I communicate with the server. I still can't start coredns on port :53 but I don't think thats a major issue. Please let me know if this is what I am supposed to get. 

![Screenshot from 2020-05-21 14-58-32](https://user-images.githubusercontent.com/51071680/82545343-32dbd200-9b74-11ea-8580-eec19ad04ec6.png)

**Comment by Chanakya-Ekbote** [comment]
![Screenshot from 2020-05-21 14-58-32](https://user-images.githubusercontent.com/51071680/82545343-32dbd200-9b74-11ea-8580-eec19ad04ec6.png)

**Comment by yongtang** [comment]
@Chanakya-Ekbote I think this is good enough. Though you want to modify the Corefile with
```
.:<another port like 9053> {
  whoami
}
```

And remove '-dns.port=1053' from command line, so
```
./coredns
```

Then you can test dig against `<another port like 9053>`. This is just to make sure `Corefile` is working correctly.

Once we can verify this, we will see if we can place an external plugin into the build.

**Comment by cekbote** [comment]
@yongtang I don't think it works. It keeps giving a repetative error bind permission denied. When I try to use sudo it says port 53 always occupied. 

Here is the bind permission error:
![bind_permission](https://user-images.githubusercontent.com/51071680/82581394-24121100-9bae-11ea-986f-a4940328ac72.png)

Here is the corefile:
![Corefile](https://user-images.githubusercontent.com/51071680/82581418-2bd1b580-9bae-11ea-9474-2c275266814a.png)

I tried to kill the process but it seems to be a system process:
![Bind_53](https://user-images.githubusercontent.com/51071680/82581447-37bd7780-9bae-11ea-90d5-8fd0091e0ede.png)

**Comment by cekbote** [comment]
@yongtang I checked coredns documentation the corefile.5.md seems to be the correct corefile. The documentation can be found here: https://github.com/coredns/coredns/blob/master/corefile.5.md

Is there something that I am doing wrong here? Also would there be a possibility that we are using different versions of coredns ?

**Comment by yongtang** [comment]
@Chanakya-Ekbote `corefile.5.md` is the document for `Corefile`. Instead you should create a file named `Corefile` and place the following inside the `Corefile`:
```
.:9053 {
  whoami
}
```

**Comment by cekbote** [comment]
> @Chanakya-Ekbote `corefile.5.md` is the document for `Corefile`. Instead you should create a file named `Corefile` and place the following inside the `Corefile`:
> 
> ```
> .:9053 {
>   whoami
> }
> ```

Ill do that and check, give me 10 min.

**Comment by cekbote** [comment]
@yongtang Yes it finally works :), thanks a lot :+1: 

I think the documentation under corefile should be edited a little bit, also maybe in the introductory readme to coredns... this is just a suggestion.  I could do it, I think, just add a few extra steps that explicitly state that we have to create a Corefile.

**Comment by yongtang** [comment]
@Chanakya-Ekbote Glad that works.

> I think the documentation under corefile should be edited a little bit, also maybe in the introductory readme to coredns... this is just a suggestion.

Would you like to create a PR to https://github.com/coredns/coredns repo? That will be a good way to get involvement in coredns community.

**Comment by cekbote** [comment]
> 
> 
> @Chanakya-Ekbote Glad that works.
> 
> > I think the documentation under corefile should be edited a little bit, also maybe in the introductory readme to coredns... this is just a suggestion.
> 
> Would you like to create a PR to https://github.com/coredns/coredns repo? That will be a good way to get involvement in coredns community.

@yongtang I'd love to. I'll make the necessary changes. :). Also please let me know how should we proceed from now on?

**Comment by yongtang** [comment]
@Chanakya-Ekbote The next step is to add an external plugin (example/demo) into your coredns build. Once you can build an external plugin and works, we will be very close to get a round trip plumbing in place.

**Comment by cekbote** [comment]
> 
> 
> @Chanakya-Ekbote The next step is to add an external plugin (example/demo) into your coredns build. Once you can build an external plugin and works, we will be very close to get a round trip plumbing in place.

@yongtang Yes Il do that too. I think I'll start with the example plugin. Thats because as you saw earlier the demo plugin was giving some errors. 

One minor update, coredns works even with Ubuntu bash on windows, I guess all we had to do was create a Corefile. :)

**Comment by yongtang** [comment]
@Chanakya-Ekbote coredns itself is cross platform and could run in windows, linux, macOS (and arm and mips). Though as you might already notice, the challenge is about golang setup. If you have a proven golang environment setup in Windows then you can just use Windows natively for coredns.

**Comment by cekbote** [comment]
> 
> 
> @Chanakya-Ekbote coredns itself is cross platform and could run in windows, linux, macOS (and arm and mips). Though as you might already notice, the challenge is about golang setup. If you have a proven golang environment setup in Windows then you can just use Windows natively for coredns.

@yongtang Oh that makes sense. 

A few updates, the example plugin does not seem to be doing what it's intended to do. Like I don't see "example" anywhere.
I tried some other plugins such as forward and log, and they seem to be working fine
I have changed the plugin.cfg file and I do have the example plugin shown when we do check or plugins. 

The Corefile

![Corefile](https://user-images.githubusercontent.com/51071680/82587916-9f2bf500-9bb7-11ea-9550-1b4bfdd21519.PNG)

The Plugin.cfg

![plugin_cfg](https://user-images.githubusercontent.com/51071680/82587951-af43d480-9bb7-11ea-9284-9145a5873a39.PNG)

The example plugin.

![example](https://user-images.githubusercontent.com/51071680/82588000-c2ef3b00-9bb7-11ea-8c59-b87211220ec6.PNG)

Port 1054 contains log and forward so you can see that works:

![forwarding_logging](https://user-images.githubusercontent.com/51071680/82588075-ddc1af80-9bb7-11ea-98a2-a600bdf985cb.PNG)

However, Port 1056 contains the example plugin.

![example_query](https://user-images.githubusercontent.com/51071680/82588192-09449a00-9bb8-11ea-805b-dbc1372003af.PNG)

Maybe you could suggest a different external plugin for testing purposes other than demo and example?

**Comment by yongtang** [comment]
@Chanakya-Ekbote demo plugin has been updated with a commit now:
https://github.com/coredns/demo/commit/57212712287e5d630b3dee70a49493b21d0e3bd0

It should work with following:
```
git clone https://github.com/coredns/coredns.git
cd coredns/plugin/
git clone https://github.com/coredns/demo.git
cd ..
vi plugin.cfg # add demo:demo to the first line 
docker run --rm -i -t -v $PWD:/v -w /v golang:1.14 sh -c 'make gen && make'
```

demo directory should be placed inside `coredns/plugin`, and your plugin.cfg should include 
```
demo:demo
```

as the first line.

**Comment by cekbote** [comment]
@yongtang  I tried it. It works now :)

![Plugin](https://user-images.githubusercontent.com/51071680/82657302-c4197a00-9c42-11ea-9a5c-c4c7426a4591.PNG)

My guess is you'd want me to integrate the go program that I created before into the demo plugin and see how that communicates with the flask server?

**Comment by yongtang** [comment]
@Chanakya-Ekbote That is great!

Yes the next step is to integrate the go program for http => flask, so that any DNS query will send a http request to flask and receive a response. Depending on the response to decide what to reply for DNS query. once this is done our end-to-end pipeline is in place and we could move to model building.

**Comment by cekbote** [comment]
@yongtang Sure, I'll do that. I'll also improve the documentation regarding the Corefile over the weekend. :)

**Comment by cekbote** [comment]
@yongtang The program is communicating with the Flask server. All it does currently is send a json object to the Flask server and gets a modified json object back. 

It can be seen here: 
![plugin_bah](https://user-images.githubusercontent.com/51071680/82727212-c80ed000-9d06-11ea-8cce-c85dfca3238d.PNG)

The modified demo plugin can be found here: 
![modified_demo](https://user-images.githubusercontent.com/51071680/82727300-42d7eb00-9d07-11ea-8800-80698d2db172.PNG)


The code for the flask server can be found here: 
![python_code](https://user-images.githubusercontent.com/51071680/82727305-47040880-9d07-11ea-83ba-f52748ca3428.PNG)

What modifications would you suggest to the reply part of the demo plugin? As I remember from the Gitter conversations we'd have to send a "blackhole IP" to the client?  From I understand if its a malicious domain, we'd be adding a "blackhole ip" to the ```rr.A = net.ParseIP(reply).To4()```  where reply would be a string that consists of a blackhole ip. Is my understanding correct?

**Comment by cekbote** [comment]
@yongtang I added a conditional in the code. The condition is that if the client wants to resolve www.google.com, it gets back 8.8.8.8 but for any other domain name, it gets back 2.2.2.2. Would that be sufficient for a barebones pipeline?

**Comment by yongtang** [comment]
@Chanakya-Ekbote For any other domains other than the one that `hit` the ml algorithm, you want to leave it alone and ask the next plugin to process. So you should use :
```
if ... {
        .... ml logic...
        // write response
	w.WriteMsg(m)
	return dns.RcodeSuccess, nil
}
// fall back to the next plugin
return plugin.NextOrFailure(n.Name(), n.Next, ctx, w, r)
```

**Comment by cekbote** [comment]
@yongtang So what I've done is if the ml algorithm predicts its a malicious domain it will return a predefined reply else, it falls back to the next plugin. To test this out should I chain a plugin after the demo plugin, maybe whoami ?

**Comment by greenpau** [comment]
> So what I've done is if the ml algorithm predicts its a malicious domain it will return a predefined reply else

@Chanakya-Ekbote , you probably want to have a configuration that allows a user to specify the address to redirect requests to malicious domains to.

That IP address could be a honeypot. When traffic comes to the address the honeypot knows there is something going on.

**Comment by cekbote** [comment]
> 
> 
> > So what I've done is if the ml algorithm predicts its a malicious domain it will return a predefined reply else
> 
> @Chanakya-Ekbote , you probably want to have a configuration that allows a user to specify the address to redirect requests to malicious domains to.
> 
> That IP address could be a honeypot. When traffic comes to the address the honeypot knows there is something going on.

I dont have any idea about honeypots. I'll read about them and we could discuss it tomorrow?

**Comment by yongtang** [comment]
@Chanakya-Ekbote honeypots is essentially a technique in security to lure and attract malicious traffic for further analysis, e.g., have a sandbox or vm to actually run malicious program for behavior analysis (and some times automatic signature generation). The honeypots itself is not the focus of this GSoC. However, you do need a way to allow user to selectively specify what the `blackhole` or `honeypots` would be when they launch the CoreDNS-ML firewall.

You can either add the option in setup.go, or even simpler, just return the IP of honeypots or blackhole in your python program in restful response.

**Comment by yongtang** [comment]
> To test this out should I chain a plugin after the demo plugin, maybe whoami ?

@Chanakya-Ekbote Ideally the coredns-ml setup should be that,
1. User setup CoreDNS the normal way with any plugins that they use (e.g., forward to 8.8.8.8, Kubernetes to wire up the Kubernetes cluster, etc).
2. User will update the Corefile to enable the ML firewall, so that ML firewall will be at the beginning of the plugin chain, and block traffic if deemed malicious. If traffic is considered ok then it will fallback to the next plugin user setup, and the rest will behave like before.

For testing purposes, you can use forward to 8.8.8.8, or whomi will be fine as well (for just testing).

**Comment by cekbote** [comment]
> 
> 
> @Chanakya-Ekbote honeypots is essentially a technique in security to lure and attract malicious traffic for further analysis, e.g., have a sandbox or vm to actually run malicious program for behavior analysis (and some times automatic signature generation). The honeypots itself is not the focus of this GSoC. However, you do need a way to allow user to selectively specify what the `blackhole` or `honeypots` would be when they launch the CoreDNS-ML firewall.
> 
> You can either add the option in setup.go, or even simpler, just return the IP of honeypots or blackhole in your python program in restful response.

@yongtang Thank you for the explanation. :)

If the barebones pipeline is ready maybe we could focus on the model creation a little bit and then come back to this?

**Comment by yongtang** [comment]
@Chanakya-Ekbote Yes I think that makes sense. We can come back and have the additional configuration options later.

**Comment by cekbote** [comment]
> 
> 
> > To test this out should I chain a plugin after the demo plugin, maybe whoami ?
> 
> @Chanakya-Ekbote Ideally the coredns-ml setup should be that,
> 
>     1. User setup CoreDNS the normal way with any plugins that they use (e.g., forward to 8.8.8.8, Kubernetes to wire up the Kubernetes cluster, etc).
> 
>     2. User will update the Corefile to enable the ML firewall, so that ML firewall will be at the beginning of the plugin chain, and block traffic if deemed malicious. If traffic is considered ok then it will fallback to the next plugin user setup, and the rest will behave like before.
> 
> 
> For testing purposes, you can use forward to 8.8.8.8, or whomi will be fine as well (for just testing).

@yongtang So what I understand is that if there is a fallthrough i.e. return Plugin.NextOrFailure , and in the Corefile lets assume the plugins are chained as follows:

.1053 {
forward . 8.8.8.8
whoami
}

Then the forward plugin will be executed first and then the whoami plugin. That's what I understood from the docs as well as your explanation. 

So what I did was I chained the ml (demo) plugin and the whoami plugin, and I was expecting both to work however only the mlplugin worked. I could not see the output of the whoami plugin. 

The corefile is as given below:
![Corefile](https://user-images.githubusercontent.com/51071680/82925677-abd79100-9f9b-11ea-9620-c3a0b5c3406b.PNG)

The code can be found here:
![Code_1](https://user-images.githubusercontent.com/51071680/82925705-b5f98f80-9f9b-11ea-8e80-48b3f5576f53.PNG)
![Code_2](https://user-images.githubusercontent.com/51071680/82925711-b7c35300-9f9b-11ea-982f-308bc627b105.PNG)

So is there some gap in my understanding or am I missing some steps?

**Comment by greenpau** [comment]
> You can either add the option in setup.go, or even simpler, just return the IP of honeypots or blackhole in your python program in restful response.

@yongtang , you are awesome at synthesizing! :+1:
